capstone/src/test/java/com/google/step/clients/MetricsClientTests.java

```diff
@@ -0,0 +1,351 @@
+
+import com.google.appengine.api.datastore.*;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.api.datastore.Query.CompositeFilter;
+import java.util.*;
+import static org.junit.Assert.*;
+import org.junit.*;
+
+
+import com.google.appengine.tools.development.testing.LocalDatastoreServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+
+import java.util.Calendar;
+
+public class MetricsClientTests {
+
+    private final LocalServiceTestHelper helper = new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
+    private final DatastoreService datastoreService = DatastoreServiceFactory.getDatastoreService();
+    private final MetricsClient metricsClient = new MetricsClient();
+
+    @Before
+    public void setUp() {
+        helper.setUp();
+    }
+
+    @After
+    public void tearDown() {
+        helper.tearDown();
+    }
+
+    public void checkIfDataStoreWorks() {
+        assertEquals(0, datastoreService.prepare(new Query("User")).countEntities(withLimit(10)));
+        datastoreService.put(new Entity("User"));
+        datastoreService.put(new Entity("User"));
+        assertEquals(2, datastoreService.prepare(new Query("User")).countEntities(withLimit(10)));
+    }
+
+    @Test
+    public void testCheckWorks1() {
+        checkIfDataStoreWorks();
+    }
+
+    @Test
+    public void testCheckWorks2() {
+        checkIfDataStoreWorks();
+    }
+
+    @Test
+    public void testPutPageView() {
+        // Checks if the putPageView method is correctly putting data into dataStore.
+        metricsClient.putPageView("Mc Donald's", "1");
+        Calendar calendar = Calendar.getInstance();
+        int year = calendar.get(Calendar.YEAR);
+        int week = calendar.get(Calendar.WEEK_OF_YEAR);
+
+        CompositeFilter compositeFilter = new Query.CompositeFilter(Query.CompositeFilterOperator.AND,
+                Arrays.asList(
+                        new FilterPredicate("restaurantKey", Query.FilterOperator.EQUAL, "1"),
+                        new FilterPredicate("year", Query.FilterOperator.EQUAL, year),
+                        new FilterPredicate("week", Query.FilterOperator.EQUAL, week)
+                ));
+
+        Query query = new Query("PageViews").setFilter(compositeFilter);
+        PreparedQuery results = datastoreService.prepare(query);
+        Entity resultEntity = results.asSingleEntity();
+        Assert.assertEquals("Count", (long) 1, resultEntity.getProperty("count"));
+    }
+
+    @Test
+    public void testSeveralPutPageViews() {
+        // Checks if the putPageView method is correctly putting and updating an entity in dataStore.
+        metricsClient.putPageView("Mc Donald's", "1");
+        metricsClient.putPageView("Mc Donald's", "1");
+        metricsClient.putPageView("Mc Donald's", "1");
+        Calendar calendar = Calendar.getInstance();
+        int year = calendar.get(Calendar.YEAR);
```

Comment:
Ok, it's like I called our earlier:

Consider injecting a Calendar as a constructor parameter. Then, you can mock it in tests and do more interesting things, like having it return different results for time, and validating that you are storing things correctly for multiple weeks.

Reply:
We can already do that through the tests themselves, since we can just hardcode the int value for the year or week. Isn't this better left the same way so that all our code depends on the same functionality for the calendar and the dates are consistent? The alternative could also be making your change and just having the servlet input the calendar.

Reply:
Indeed if you're doing time-sensitive things and you're validating many classes working together, then as long as things behave well, it's fine to use real time.

I won't make you change it right now, because I know there's more important things to do, but I'll still write this out, just for your edification:

I've found that it's generally a best practice to set some kind of a reference time that stays constant throughout a test, and set all timestamps on entities you set up to something that comes before that reference time. This guarantees a few things.

First, it makes it so your tests execute exactly the same way every time.

Second, you prevent weird race conditions in your tests that could result in nondeterministic behavior. Consider your specific example, actually: Theoretically, if you run your tests on Saturday night around midnight, you could get a different week when your test looks it up, and when your implementation looks it up. That would result in the test failing on "wrong data" because it was expecting this week but it got next week.

I know that won't happen (because hopefully you're not working at midnight on the weekends haha), but you can see how time-related code can result in inconsistent test behavior if not kept in check (e.g. imagine if you were looking at seconds instead of weeks - then, your continuous tests could break a few times a week).

Whenever possible/reasonable, it's good to avoid race conditions.

Again, no need to make the change here.